tlair/src/test/scala/at/forsyte/apalache/tla/lir/transformations/standard/TestDeepCopy.scala

```diff
@@ -0,0 +1,84 @@
+package at.forsyte.apalache.tla.lir.transformations.standard
+
+import at.forsyte.apalache.tla.lir.transformations.impl.IdleTracker
+import at.forsyte.apalache.tla.lir._
+import org.scalacheck.Prop.forAll
+import org.scalatest.prop.Checkers
+import org.scalatest.{BeforeAndAfter, FunSuite}
+
+/**
+ * Tests for DeepCopy using Scalacheck.
+ *
+ * @author Igor Konnov
+ */
+class TestDeepCopy extends FunSuite with BeforeAndAfter with Checkers {
+  private def equalExCopies: (TlaEx, TlaEx) => Boolean = {
+    case (l @ ValEx(lval), r @ ValEx(rval)) =>
+      l.ID != r.ID && lval == rval && l.typeTag == r.typeTag
+
+    case (l @ NameEx(lname), r @ NameEx(rname)) =>
+      l.ID != r.ID && lname == rname && l.typeTag == r.typeTag
+
+    case (l @ OperEx(lop, largs @ _*), r @ OperEx(rop, rargs @ _*)) =>
+      l.ID != r.ID &&
+        lop == rop &&
+        l.typeTag == r.typeTag &&
+        largs.length == rargs.length &&
+        largs.zip(rargs).forall(equalExCopies.tupled)
+
+    case (l @ LetInEx(lbody, ldefs @ _*), r @ LetInEx(rbody, rdefs @ _*)) =>
+      equalExCopies(lbody, rbody) &&
+        l.ID != r.ID &&
+        l.typeTag == r.typeTag &&
+        ldefs.length != rdefs.length &&
+        ldefs
+          .zip(rdefs)
+          .forall(equalDeclCopies.tupled)
+
+    case _ => false
+  }
+
+  private def equalDeclCopies: (TlaDecl, TlaDecl) => Boolean = {
+    case (l @ TlaConstDecl(lname), r @ TlaConstDecl(rname)) =>
+      l.ID != r.ID && lname == rname && l.typeTag == r.typeTag
+
+    case (l @ TlaVarDecl(lname), r @ TlaVarDecl(rname)) =>
+      l.ID != r.ID && lname == rname && l.typeTag == r.typeTag
+
+    case (l @ TlaAssumeDecl(lbody), r @ TlaAssumeDecl(rbody)) =>
+      l.ID != r.ID && equalExCopies(lbody, rbody) && l.typeTag == r.typeTag
+
+    case (l @ TlaTheoremDecl(lname, lbody), r @ TlaTheoremDecl(rname, rbody)) =>
+      l.ID != r.ID &&
+        lname == rname &&
+        equalExCopies(lbody, rbody) &&
+        l.typeTag == r.typeTag
+
+    case (l @ TlaOperDecl(lname, lparams, lbody), r @ TlaOperDecl(rname, rparams, rbody)) =>
+      l.ID != r.ID &&
+        lname == rname &&
+        lparams == rparams &&
+        equalExCopies(lbody, rbody) &&
+        l.isRecursive == r.isRecursive &&
+        l.typeTag == r.typeTag
+
+    case _ => false
+  }
+
+  test("deep copy replicates expressions and declarations but not their identifiers") {
+    val gens = new IrGenerators {
+      override val maxArgs: Int = 10
+    }
+
+    val prop = {
+      def exGen(ctx: gens.UserContext) = gens.genTlaEx(gens.simpleOperators ++ gens.setOperators)(ctx)
```

Comment:
@konnov I think you are never hitting the `LetInEx` branch because these generator sets don't include letIn expressions: https://github.com/informalsystems/apalache/blob/unstable/tlair/src/test/scala/at/forsyte/apalache/tla/lir/IrGenerators.scala#L57-L71

Reply:
True. I have added `LetInEx`. Still this case is not covered.

Reply:
OK. Now I have fixed the generator in the right place. The test is failing. However, we should introduce coverage of our tests.

Reply:
Agree!